[rpc] throw correct Exception on local client based on the RemoteException
type.

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#32936 [rpc] throw correct Exception on local client based on the RemoteException
type.**
type.**
type.**

Closes https://github.com/pytorch/pytorch/issues/32732. Currently if a UDF run in RPC throws an exception such as ValueError or TypeError, we wrap this in a RemoteException on the callee side. When raising this on the caller side, we currently raise a vanilla Exception. This diff changes it so that the correct exception is thrown. Tested by changing the current rpc tests to assert on the right type of error rather than just the base `Exception`.

Differential Revision: [D19700434](https://our.internmc.facebook.com/intern/diff/D19700434/)

rohan-varma committed:
[rpc] throw correct Exception on local client based on the RemoteException
type.

Closes https://github.com/pytorch/pytorch/issues/32732. Currently if a
UDF run in RPC throws an exception such as ValueError or TypeError, we wrap
this in a RemoteException on the callee side. When raising this on the caller
side, we currently raise a vanilla Exception. This diff changes it so that the
correct exception is thrown. Tested by changing the current rpc tests to assert
on the right type of error rather than just the base `Exception`.

Differential Revision: [D19700434](https://our.internmc.facebook.com/intern/diff/D19700434/)

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit 14cb0016:


* **3/3** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=14cb00163a1eb00004b57dbedcc4abea5a8dd198).

### :detective: 3 new failures recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4403743) pytorch_linux_xenial_py3_5_build (1/3)
**Step:** "Should run job" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=94696176) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=94696176))
<details>
<summary>
<code>/home/circleci/workspace/.circleci/scripts/should_run_job.sh: line 25: circleci: command not found</code>
</summary>

```
+ '[' -z pytorch-linux-xenial-py3.5-build ']' 
+ '[' -e /home/circleci/workspace/.circleci/scripts/COMMIT_MSG ']' 
+ '[' -n https://github.com/pytorch/pytorch/pull/32936 ']' 
+ [[ gh/rohan-varma/75/head != \c\i\-\a\l\l\/* ]] 
+ [[ gh/rohan-varma/75/head != \n\i\g\h\t\l\y ]] 
+ [[ gh/rohan-varma/75/head != \p\o\s\t\n\i\g\h\t\l\y ]] 
+ '[' -e /home/circleci/workspace/.circleci/scripts/should_run_job.py ']' 
+ python /home/circleci/workspace/.circleci/scripts/should_run_job.py pytorch-linux-xenial-py3.5-build 
Rejecting pytorch-linux-xenial-py3.5-build 
+ circleci step halt 
/home/circleci/workspace/.circleci/scripts/should_run_job.sh: line 25: circleci: command not found 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4403757) pytorch_parallelnative_linux_xenial_py3_6_gcc5_4_build (2/3)
**Step:** "Should run job" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=94696177) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=94696177))
<details>
<summary>
<code>/home/circleci/workspace/.circleci/scripts/should_run_job.sh: line 25: circleci: command not found</code>
</summary>

```
+ '[' -z pytorch-parallelnative-linux-xenial-py3.6-gcc5.4-build ']' 
+ '[' -e /home/circleci/workspace/.circleci/scripts/COMMIT_MSG ']' 
+ '[' -n https://github.com/pytorch/pytorch/pull/32936 ']' 
+ [[ gh/rohan-varma/75/head != \c\i\-\a\l\l\/* ]] 
+ [[ gh/rohan-varma/75/head != \n\i\g\h\t\l\y ]] 
+ [[ gh/rohan-varma/75/head != \p\o\s\t\n\i\g\h\t\l\y ]] 
+ '[' -e /home/circleci/workspace/.circleci/scripts/should_run_job.py ']' 
+ python /home/circleci/workspace/.circleci/scripts/should_run_job.py pytorch-parallelnative-linux-xenial-py3.6-gcc5.4-build 
Rejecting pytorch-parallelnative-linux-xenial-py3.6-gcc5.4-build 
+ circleci step halt 
/home/circleci/workspace/.circleci/scripts/should_run_job.sh: line 25: circleci: command not found 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4403754) pytorch_paralleltbb_linux_xenial_py3_6_gcc5_4_build (3/3)
**Step:** "Should run job" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=94696365) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=94696365))
<details>
<summary>
<code>/home/circleci/workspace/.circleci/scripts/should_run_job.sh: line 25: circleci: command not found</code>
</summary>

```
+ '[' -z pytorch-paralleltbb-linux-xenial-py3.6-gcc5.4-build ']' 
+ '[' -e /home/circleci/workspace/.circleci/scripts/COMMIT_MSG ']' 
+ '[' -n https://github.com/pytorch/pytorch/pull/32936 ']' 
+ [[ gh/rohan-varma/75/head != \c\i\-\a\l\l\/* ]] 
+ [[ gh/rohan-varma/75/head != \n\i\g\h\t\l\y ]] 
+ [[ gh/rohan-varma/75/head != \p\o\s\t\n\i\g\h\t\l\y ]] 
+ '[' -e /home/circleci/workspace/.circleci/scripts/should_run_job.py ']' 
+ python /home/circleci/workspace/.circleci/scripts/should_run_job.py pytorch-paralleltbb-linux-xenial-py3.6-gcc5.4-build 
Rejecting pytorch-paralleltbb-linux-xenial-py3.6-gcc5.4-build 
+ circleci step halt 
/home/circleci/workspace/.circleci/scripts/should_run_job.sh: line 25: circleci: command not found 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

xush6528 commented on torch/distributed/rpc/internal.py:
Like this change!

rohan-varma committed:
Update on "[rpc] throw correct Exception on local client based on the RemoteException
type."

type.**

Closes https://github.com/pytorch/pytorch/issues/32732. Currently if a UDF run in RPC throws an exception such as ValueError or TypeError, we wrap this in a RemoteException on the callee side. When raising this on the caller side, we currently raise a vanilla Exception. This diff changes it so that the correct exception is thrown. Tested by changing the current rpc tests to assert on the right type of error rather than just the base `Exception`.

Differential Revision: [D19700434](https://our.internmc.facebook.com/intern/diff/D19700434/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[rpc] throw correct Exception on local client based on the RemoteException
type."

type.**
type.**

Closes https://github.com/pytorch/pytorch/issues/32732. Currently if a UDF run in RPC throws an exception such as ValueError or TypeError, we wrap this in a RemoteException on the callee side. When raising this on the caller side, we currently raise a vanilla Exception. This diff changes it so that the correct exception is thrown. Tested by changing the current rpc tests to assert on the right type of error rather than just the base `Exception`.

Differential Revision: [D19700434](https://our.internmc.facebook.com/intern/diff/D19700434/)

[ghstack-poisoned]

rohan-varma commented:
The lint issue is due to an untouched file.
